### PR TITLE
executor: tiny refactor the Executor interface

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

There is trending that everything is put into an interface, make the interface fat, less composable.

I don't think those methods should be part of the `Executor` interface.

```
	retTypes() []*types.FieldType
	newFirstChunk() *chunk.Chunk
```

### What is changed and how it works?

Introduce a This() method to get the baseExecutor, so we don't need to add
the method from baseExecutor to the Executor interface any more.

Before:
```
type Executor interface {
	Open(context.Context) error
	Next(ctx context.Context, req *chunk.RecordBatch) error
	Close() error
	Schema() *expression.Schema

	retTypes() []*types.FieldType
	newFirstChunk() *chunk.Chunk
}
```

After:
```
type Executor interface {
        This() *baseExecutor
	Open(context.Context) error
	Next(ctx context.Context, req *chunk.RecordBatch) error
	Close() error
	Schema() *expression.Schema
}
```

`retTypes` and `newFirstChunk` are also changed to functions accordingly.

### Check List 

Tests 

 - No code

Side effects

 - Possible performance regression
 - Breaking backward compatibility

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @coocood @lysu

**Comment by shenli** [comment]
Definitely, this PR is not a "No Code" PR.

File: executor/executor.go [hunk 1]
```diff
@@ -166,13 +173,14 @@ func newBaseExecutor(ctx sessionctx.Context, schema *expression.Schema, id fmt.S
 // return a batch of rows, other than a single row in Volcano.
 // NOTE: Executors must call "chk.Reset()" before appending their results to it.
 type Executor interface {
+	This() *baseExecutor
```
Comment (coocood) [inline executor/executor.go]: `This` can be unexported, and why not name it `base`?

Comment (zz-jason) [inline executor/executor.go]: In the OOP concept, `This` is commonly used to represent the actual object pointer.

File: executor/executor.go [hunk 2]
```diff
@@ -166,13 +173,14 @@ func newBaseExecutor(ctx sessionctx.Context, schema *expression.Schema, id fmt.S
...
 	Open(context.Context) error
 	Next(ctx context.Context, req *chunk.RecordBatch) error
 	Close() error
 	Schema() *expression.Schema
 
-	retTypes() []*types.FieldType
-	newFirstChunk() *chunk.Chunk
+	// retTypes() []*types.FieldType
```
Comment (zz-jason) [inline executor/executor.go]: How about removing them from the code instead of commenting them out?

**Comment by tiancaiamao** [comment]
There is no logic change, just a simple refactor, I don't introduce new code in this PR. @shenli

File: executor/executor.go [hunk 3]
```diff
@@ -117,13 +122,15 @@ func (e *baseExecutor) Schema() *expression.Schema {
 }
 
 // newFirstChunk creates a new chunk to buffer current executor's result.
-func (e *baseExecutor) newFirstChunk() *chunk.Chunk {
-	return chunk.New(e.retTypes(), e.initCap, e.maxChunkSize)
+func newFirstChunk(e Executor) *chunk.Chunk {
+	base := e.This()
+	return chunk.New(base.retFieldTypes, base.initCap, base.maxChunkSize)
```
Comment (zz-jason) [inline executor/executor.go]: @qw4990 @lysu I think we don't need initial capacity anymore? Can we remove this mechanism?

File: executor/point_get.go [hunk 4]
```diff
@@ -242,21 +245,21 @@ func getColInfoByID(tbl *model.TableInfo, colID int64) *model.ColumnInfo {
...
+// 		}
+// 	}
+// 	return e.tps
+// }
 
-func (e *PointGetExecutor) newFirstChunk() *chunk.Chunk {
-	return chunk.New(e.retTypes(), 1, 1)
-}
+// func (e *PointGetExecutor) newFirstChunk() *chunk.Chunk {
+// 	return chunk.New(e.retTypes(), 1, 1)
```
Comment (coocood) [inline executor/point_get.go]: General `newFirstChunk` don't use 1, 1 as cap and max chunk size, which wastes memory.

**Comment by zz-jason** [comment]
> There is no logic change, just a simple refactor, I don't introduce new code in this PR. @shenli

Function `newFirstChunk()` and `retTypes()` are new code :joy:

File: executor/executor.go (hunk 1 above)
Comment (tiancaiamao) [reply executor/executor.go]: Done

Comment (tiancaiamao) [reply executor/executor.go]: It's changed to `base` by the suggestion from @coocood

File: executor/executor.go (hunk 2 above)
Comment (tiancaiamao) [reply executor/executor.go]: Done

File: executor/point_get.go [hunk 5]
```diff
@@ -19,10 +19,10 @@ import (
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/parser/model"
 	"github.com/pingcap/parser/mysql"
-	"github.com/pingcap/tidb/expression"
+	// "github.com/pingcap/tidb/expression"
```
Comment (zz-jason) [inline executor/point_get.go]: should be removed

File: executor/point_get.go [hunk 6]
```diff
@@ -19,10 +19,10 @@ import (
...
 	"github.com/pingcap/tidb/kv"
 	plannercore "github.com/pingcap/tidb/planner/core"
-	"github.com/pingcap/tidb/sessionctx"
+	// "github.com/pingcap/tidb/sessionctx"
```
Comment (zz-jason) [inline executor/point_get.go]: ditto

File: executor/point_get.go (hunk 4 above)
Comment (tiancaiamao) [reply executor/point_get.go]: Fixed

File: executor/executor.go (hunk 1 above)
Comment (zz-jason) [reply executor/executor.go]: OK. But I'm wondering whether this PR is beneficial. 

This PR removed two functions from the `Executor` interface:
* `retTypes() []*types.FieldType`
* `newFirstChunk() *chunk.Chunk`

But it also introduced one function to the `Executor` interface:
* `base() *baseExecutor`

So actually, we removed 1 function indeed.

Comment (zz-jason) [reply executor/executor.go]: I think `base()` should not be part of the Executor interface either.

File: executor/point_get.go (hunk 5 above)
Comment (tiancaiamao) [reply executor/point_get.go]: Done

**Comment by coocood** [comment]
LGTM

File: executor/executor.go (hunk 1 above)
Comment (zz-jason) [reply executor/executor.go]: But in reality, we have to populate some help methods to the `Executor` interface. Not just TiDB, but also other database systems. For example:
* Impala: https://github.com/cloudera/Impala/blob/cdh5-trunk/be/src/exec/exec-node.h

File: executor/executor.go (hunk 3 above)
Comment (lysu) [reply executor/executor.go]: yes

**Pull request merged**